### ISSUE TITLE
Unable to read data from 16f877a

### ISSUE THREAD

**Issue by hernandik** [makerprojects/ProgramPic#4]
I am unable to read data from the 16f877a model.
Or it read 0x3fff or 0x00ff.
However, it can get correctly the configs from the device. It seens to reads fine the device ID, 0x0e2x
I'm reading the code, but was unable to spot any problem.
I'm using a driver for the 13V line. It seens to work properly.
I use a driver for the +5v line. I also use direct connection with the arduino power control pin.
No success. Some sugestion about that?

**Comment by makerprojects** [comment]
Reading 0x3ff might indicate a not programmed controller.

 

My suggestion would be to start with trying to read another controller which you know for sure to be programmed to check first whether the reading as such is working. 

 

 

Von: Hernandi F. Krammes F. [mailto:notifications@github.com] 
Gesendet: Dienstag, 16. Januar 2018 23:16
An: makerprojects/ProgramPic
Cc: Subscribed
Betreff: [makerprojects/ProgramPic] Unable to read data from 16f877a (#4)

 

I am unable to read data from the 16f877a model.
Or it read 0x3fff or 0x00ff.
However, it can get correctly the configs from the device. It seens to reads fine the device ID, 0x0e2x
I'm reading the code, but was unable to spot any problem.
I'm using a driver for the 13V line. It seens to work properly.
I use a driver for the +5v line. I also use direct connection with the arduino power control pin.
No success. Some sugestion about that?

**Comment by hernandik** [comment]
The PIC controller had a program.
It was not locked.
I had programmed it in the past.
I will test the controller to be sure that it is still running.
I monitored the serial data during a read process, and the PIC returns only 0xffff in the line bus.
I reduce the speed of the bitbang to watch if the cable impendance wasn't causing any issue.
I'll test and put a answer about this.

**Comment by hernandik** [comment]
Hi. Up to now, I tested again the programmer. I inserted a 16f877a with a know program and was able to read the program. I tried to burn a new program inside de device and after that I was unable to read the correct program again. It seens that the problem is related to the programming routines for the 16f877a. I'm debugging now.

**Comment by hernandik** [comment]
I found the problem and fixed it. My pics became locked because of the incorrect config word being programmed. The routine to write the config word has little differences from the other pics. I'm posting a fix in my repo.

**Comment by hernandik** [comment]
Also seen like the erase command need a longer time to work
DELAY_TFULLERA 50000 ->DELAY_TFULLERA 10000

**Comment by eslammohamedahmed** [comment]
Hello @hernandik ,
i have the same problem at read/write but my pic16f877a is new i try 2 PICs but the same problem at every time can you help?
note: this is my personal edited sketch i add pic16f877a to it pls. can look to it?
[ProgramPic.txt](https://github.com/makerprojects/ProgramPic/files/1671123/ProgramPic.txt)
regards and thanks at all.

**Comment by hernandik** [comment]
Take a look at my branch. I post on it the fix. However i didn't test it on other pics; i believe that the change can affect other pics models.

**Comment by hernandik** [comment]
I really don't know if I should share the fix in this branch.
In my branch, I reformated the entire code because it was poorly formatted. https://github.com/hernandik/ProgramPic

**Comment by makerprojects** [comment]
Let me take a look at the changes you made this week and then we can discuss. Thx. 

 

Von: Hernandi F. Krammes F. [mailto:notifications@github.com] 
Gesendet: Sonntag, 28. Januar 2018 17:14
An: makerprojects/ProgramPic
Cc: makerprojects; Comment
Betreff: Re: [makerprojects/ProgramPic] Unable to read data from 16f877a (#4)

 

I really don't know if I should share the fix in this branch.
In my branch, I reformated the entire code because it was poorly formatted. https://github.com/hernandik/ProgramPic

**Comment by makerprojects** [comment]
I took a look at the changes you made. The source code is nicely formatted but I am not sure that these changes have no impact on programming other controllers. Also, the changes are not fully compatible with the (optional) GUI provided. Meanwhile I ordered a 16f877a for my own testing which I will do in the next couple of weeks. I will remove the initial / current implementation of supporting the 16f877a in my repository since it is not working anyway and test / verify the proposed changes.
